### PR TITLE
topk_large_indices: fused segmented body for wide rows at K 512

### LINKED ISSUE #55763: topk_large_indices charges every k below 512 the full 512 bucket sort price; above 32 chunks k=1024 is faster than any k <= 512
from a topk_large_indices perf sweep on p100a at main tip 2dbd14bf632 (2026-09-12, program cache on, tracy device kernel duration, medians of 3 timed calls after 3 discarded, bf16 row major DRAM input; internal ref TEN-4859). body rewritten september 15: this replaces the numbers taken at the July kernel state, which predate the fused bodies of PR 53953 (merged 2026-08-27).

the op snaps the requested k up to a sort bucket K' in {512, 1024, 2048} (topk_large_indices_program_factory.cpp:34-41; the compute kernel static_asserts those three, compute.cpp:48) and picks a body mode from the bucket and the chunk count (:130-142): FusedEndToEnd at K' 512 with up to 32 chunks, Classic at K' 512 above that, FusedSegmented at K' 1024 and 2048. the compute cost is set by the bucket and the mode, not by k: k = 16, 32, 64, 256 and 512 measure the same duration within 0.5 percent at every N >= 1024 (one row: 51.7 to 51.9 us at N=16384, 556.0 us at N=131072; below N=1024 a single chunk row is a 3 us kernel and its call to call jitter reaches 2 percent). per row on one core, t_row = c0(K') + a_tot(K') x ceil(N / K') + c_seg(K') x (segments - 1), with a_tot 2,123 cycles per 512 chunk fused end to end (4.15 cycles per scanned element), 2,921 classic (5.71), 4,901 per 1024 chunk (4.79) and 11,436 per 2048 chunk (5.58); c0 1,969 / 4,012 / 6,300 cycles at K' 512 / 1024 / 2048; c_seg 2,393 / 5,677 per extra 32 chunk segment at K' 1024 / 2048. the fused bodies are 25 to 28 percent cheaper per chunk than the July kernel at the price of 1.5 to 2.9x larger intercepts; the nightly pin (640, 51200, 1536) reads 1,284.7 us against 1,684.0 then. the op stays compute bound on the SFPU sort: an L1 resident input runs at 0.91 to 0.99 of the DRAM duration at k 2048 (N 4096 to 65536, 110 rows).

so a small k selection over a long row still pays the K' 512 sort per chunk, and since PR 53464 (merged 2026-08-19) more callers land here: ttnn.topk now routes its large k, wide non pow2 small k and MoE gate shapes onto this op on blackhole. the minimax MSA block select (640 rows of 440 block scores, k 16) is one 512 wide chunk per row, 6 waves over 110 cores, 14.3 us, most of it the per op intercept; the deepseek indexer shapes (2048 rows, N 8192, k 1024 / 2048) measure 567.0 / 670.1 us.

one new consequence of the body selection: the 512 bucket has no fused segmented body, so above 32 chunks it falls back to the classic body, and at N=131072 a request for any k <= 512 (556.0 us) is 15 percent slower than k=1024 (472.4 us) and 1 percent slower than k=2048 (550.8 us). callers with small k over rows longer than 16384 currently do best by asking for k=1024. the ask stands: a narrower sort bucket or a cheaper small k body, and a segmented fused body for K'=512. the open PR 53457 carries a data dependent chunk skip that cuts work for small k on rows whose later chunks cannot rank, but no narrower bucket; PR 53457 and PR 53989 (column split) change how rows are split over cores, not the bucket.

repro: analysis/campaigns/topk_campaign.py --classes L1,L2,L3,L4,L5 on branch mvlahovic/sdpa_topk_harness (analysis/RUNBOOK.md section 7); ttnn.experimental.topk_large_indices on [1, 1, R, N] bf16 row major with k in {16..512} at fixed N reproduces the flatness.

### PR DESCRIPTION
**Pull request by mvlahovicTT**
`topk_large_indices` snaps k to a 512, 1024 or 2048 wide sort. Rows wider than 32 chunks at the 512 width ran the classic body, which re inits, splits and merges per chunk, so a k of 64 over 131072 elements cost more per element than the same row at k 1024. The segmented fused body already handles any width at 1024 and 2048; this uses it for the 512 width too. Per chunk cost goes from about 2,900 to about 2,100 cycles.

`topk_large_indices` device kernel duration on a p100a, median of 30 calls after 3 warmups:

| rows, N, k | before | after |
|---|---|---|
| 1, 131072, 64 | 555.7 us | 411.3 us (-26%) |
| 1, 131072, 256 | 555.8 us | 411.3 us (-26%) |
| 1, 131072, 512 | 555.8 us | 411.4 us (-26%) |
| 1, 131072, 1024 | 472.3 us | 472.3 us (+0%) |
| 32, 131072, 512 | 556.0 us | 411.6 us (-26%) |
| 32, 65536, 64 | 278.5 us | 206.3 us (-26%) |

Tests: the fused segment boundary cases now include 64, 65, 129, 256 and 257 chunks at k 512, a spread winners case at 40 chunks, and the program cache test names the two fused bodies. `test_topk_large_indices.py` 80 passed, routed `test_topk.py` cases 36 passed on the p100a.

Fixes #55763

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mvlahovic/topk-large-indices-segmented-512)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/topk_large_indices/device/topk_large_indices_program_factory.cpp [hunk 1]
```diff
@@ -115,16 +115,17 @@ void set_runtime_args(
...
+    // public k values in [528, 1008] get the same fused body as k=1024.
     if (llk_k >= to_uint32(LlkTargetK::K1024)) {
         return ComputeBodyMode::FusedSegmented;
     }
 
+    // The classic body re-inits, splits and merges per chunk, so a wide row at K 512 used to
+    // cost more per element than the same row at K 1024. Segmented fusion is cheaper past 32 chunks too.
     const uint32_t physical_chunks = tt::div_up(input_last_dim, llk_k);
-    return physical_chunks <= 32 ? ComputeBodyMode::FusedEndToEnd : ComputeBodyMode::Classic;
+    return physical_chunks <= 32 ? ComputeBodyMode::FusedEndToEnd : ComputeBodyMode::FusedSegmented;
```
[the remark this answers is not in the document]
Comment (mvlahovicTT) [reply ttnn/cpp/ttnn/operations/experimental/topk_large_indices/device/topk_large_indices_program_factory.cpp:128]: removed the classic body and the enum value, nothing selected it any more (ec84af1c6ad, 151 topk_large_indices tests pass).

[the remark this answers is not in the document]
Comment (mvlahovicTT) [reply ttnn/cpp/ttnn/operations/experimental/topk_large_indices/device/topk_large_indices_program_factory.cpp:128]: removed the classic body and the enum value, nothing selected it any more (ec84af1c6ad, 151 topk_large_indices tests pass).
